src/Microsoft.AspNetCore.HttpsPolicy/HttpsRedirectionMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _config = config ?? throw new ArgumentException(nameof(config));
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            _serverAddressesFeature = serverAddressesFeature ?? throw new ArgumentNullException(nameof(serverAddressesFeature));
+
+            var httpsRedirectionOptions = options.Value;
+            _httpsPort = httpsRedirectionOptions.HttpsPort;
+            _statusCode = httpsRedirectionOptions.RedirectStatusCode;
+        }
+
+        /// <summary>
+        /// Invokes the HttpsRedirectionMiddleware
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            if (context.Request.IsHttps)
+            {
+                return _next(context);
+            }
+
+            if (!_httpsPort.HasValue)
+            {
+                CheckForHttpsPorts();
+            }
+
+            var host = context.Request.Host;
+            if (_httpsPort != 443)
+            {
+                host = new HostString(host.Host, _httpsPort.Value);
+            }
+            else
+            {
+                host = new HostString(host.Host);
+            }
+
+            var request = context.Request;
+            var redirectUrl = UriHelper.BuildAbsolute(
+                "https", 
+                host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+
+            context.Response.StatusCode = _statusCode;
+            context.Response.Headers[HeaderNames.Location] = redirectUrl;
+
+            return Task.CompletedTask;
+        }
+
+        private void CheckForHttpsPorts()
+        {
+            // The IServerAddressesFeature will not be ready until the middleware is Invoked,
+            // Order for finding the HTTPS port:
+            // 1. Set in the HttpsRedirectionOptions
+            // 2. HTTPS_PORT environment variable
+            // 3. IServerAddressesFeature
+            // 4. 443 (or not set)
+
+            var configHttpsPort = _config.GetValue<int?>("HTTPS_PORT");
+            if (configHttpsPort.HasValue)
+            {
+                _httpsPort = configHttpsPort;
+                return;
+            } 
+
+            int? httpsPort = null;
+            foreach (var address in _serverAddressesFeature.Addresses)
+            {
+                var bindingAddress = BindingAddress.Parse(address);
+                if (bindingAddress.Scheme == "https")
+                {
+                    // If we find multiple different https ports specified, throw
+                    if (httpsPort != null && httpsPort != bindingAddress.Port)
+                    {
+                        throw new ArgumentException($"Cannot specify multiple https ports in IServerAddressesFeature. " +
```

Comment:
"Cannot determine the https port from IServerAddressesFeature, multiple values were found. Please set the desired port explicitly on HttpsRedirectionOptions.HttpsPort."